fdbserver/MasterProxyServer.actor.cpp

```diff
@@ -1972,6 +1976,85 @@ ACTOR Future<Void> proxyCheckSafeExclusion(Reference<AsyncVar<ServerDBInfo>> db,
 	return Void();
 }
 
+ACTOR Future<Version> getTxnStateLifetime(Database db) {
+	state Transaction tr(db);
+	loop {
+		try {
+			tr.setOption(FDBTransactionOptions::LOCK_AWARE);
+			Optional<Value> val = wait(tr.get(txStateLifetime));
+			if (val.present()) {
+				return BinaryReader::fromStringRef<Version>(val.get(), IncludeVersion());
+			} else {
+				tr.set(txStateLifetime, BinaryWriter::toValue(SERVER_KNOBS->TXN_STATE_LIFE_VERSIONS, IncludeVersion()));
+				wait(tr.commit());
+				return SERVER_KNOBS->TXN_STATE_LIFE_VERSIONS;
```

Comment:
This is a slightly awkward pattern for a knob, as it makes it unchangeable after a cluster is created. There are other such knobs, but I think it would be worth considering whether this is something we can make mutable, maybe by giving up on all old commits when it changes?

Reply:
You're right. I think this logic can also go into the background cleaner. In that case this config parameter can be set to `0` - but the cleaner will never delete past this window.